rfcs/rfc-metaprogramming.md

```diff
@@ -0,0 +1,379 @@
+
+Returns true if the named package exists.
+
+For example:
+
+```perl
+my $bool = meta::has_package("Example::Package");
+# behaves as
+my $bool = do { no strict 'refs'; exists ${"Example::"}{"Package::"} };
+```
+
+#### `get_package`
+
+```perl
+$metapackage = meta::get_package($name);
+```
+
+Returns a meta-package object instance to represent the named package, if such a package exists. If not an exception is thrown.
+
+#### `add_package`
+
+```perl
+$metapackage = meta::add_package($name);
+```
+
+Creates a new package of the given name, and returns a meta-package object instance to represent it.
+
+#### `list_packages`
+
+```perl
+@metapackages = meta::list_packages($basename);
+```
+
+Returns a list of sub-packages within the given base package name, each as a meta-package object instance.
+
+### Functions to operate on References
+
+Some less common functions that may still be useful in some situations. These functions can be used to obtain meta-programming object instances in ways other than walking the symbol table for given names.
+
+#### `get_variable`
+
+```perl
+$metavar = meta::get_variable($varref);
+```
+
+Given a reference to a package variable, returns a meta-variable object instance to represent it. If `$subref` is not a SCALAR, ARRAY or HASH reference, an exception is thrown.
+
+*Note:* It is currently unspecified what happens if the reference refers to a lexical or anonymous variable, rather than a package one. This RFC does not specify any useful behaviour for such instances, but it's possible that future abilities may be added that become useful on lexical variables or anonymous containers.
+
+#### `get_subroutine`
+
+```perl
+$metasub = meta::get_subroutine($subref);
+```
+
+Given a CODE reference, returns a meta-subroutine object instance to represent it. If `$subref` is not a CODE reference, an exception is thrown.
+
+This is primarily useful for working with newly-constructed anonymous functions, in order to give them a valid name by calling `set_subname` for instance.
+
+### Methods on Meta-Package Objects
+
+These methods all operate on a meta-package object obtained by one of the above functions.
+
+#### `name`
+
+```perl
+$name = $metapackage->name;
+```
+
+Returns the fully-qualified name of the package as a plain string. This will match the string passed to `meta::get_package` used to obtain it, for example.
+
+#### `has_symbol`
+
+```perl
+$bool = $metapackage->has_symbol($name);
+```
+
+Returns true if the named symbol exists. `$name` must begin with a sigil; one of the characters "$@%&". 
+
+For example:
+
+```perl
+my $bool = meta::get_package('Example::Package')->has_symbol('$OPT');
+
+# behaves as
+my $bool = do { no strict 'refs'; defined *{"Example::Package::OPT"}{SCALAR} };
+```
+
+#### `get_symbol`
+
+```perl
+$metasymbol = $metapackage->get_symbol($name);
+```
+
+Returns a meta-symbol object instance to represent the symbol within the package, if such a symbol exists. If not an exception is thrown.
```

Comment:
Hmm, I thought about that - I guess there's situations you want it either way around. Sometimes you want to optionally work on a symbol if it's there; other times you definitely expect it to be there and want an error if it isn't.

If it just silently returns undef then code like this would fail:

```perl
$metapackage->get_symbol('&func')->set_prototype('($$$)');
```

Or rather, it would fail with the rather nonobvious "Can't call method "set_prototype on an undefined value" rather than some slightly more meaningful message about whatever package name it was not containing a '&func' symbol.

Since we don't want ((I think?)) two sets of methods having "get or throw" vs "get if exists otherwise return undef" semantics, we should decide which one we want here. As you point out it's always possible to make an optional one out of a throwy one; it's also possible to make vice-versa. If the method just returned undef if the symbol doesn't exist, one could always:

```perl
($metapackage->get_symbol('&func') or die "Package does not have a &func symbol")
   ->set_prototype('($$$)');
```

Though it does involve a bit of repeating the name. Plus it feels very 1980s "return undef and set errno" technology. We have exceptions now - should we not use them?

Reply:
I could see providing both a throwing and a non-throwing method, whether a missing symbol is exceptional depends on the caller's circumstances.

I suggest different methods here rather than a flag both in having a method that "does one thing", and for better tool support, I know Coverity for example is confused by sv_utf8_downgrade() which takes a flag to decide whether to throw and complains that only some callers check the return value.

In this case the caller always *saves* the return value, but for a throwing method wouldn't need to check its truthiness.

Reply:
I could see an analogous `find_symbol('&func')` which would return `undef`, but I agree that having that and `get_symbol` side-by-side might be confusing (or clunky). If we go with exceptions, I won't cry :)